File: tree/trie/trie1.cpp
#include <vector>

using namespace std;

/*
 * 子节点用 vector 实现
 */

const int ALPHABETS = 26;

struct TrieNode
{
    bool terminal;
    vector<TrieNode*> children;
    TrieNode()
    {
        terminal = false;
        children = vector<TrieNode*>(ALPHABETS, nullptr);
    }
    ~TrieNode(){}
};

class Trie
{
public:
    Trie()
    {
        root = new TrieNode();
    }

    ~Trie()
    {
        if(root)
            delete_sub_tree(root);
        delete root;
        root = nullptr;
    }

    void insert(const string& word)
    {
        TrieNode *iter = root;
        for(const char& ch: word)
        {
            TrieNode *&nxt = ;
            if(!(iter -> children)[ch - 'a'])
                (iter -> children)[ch - 'a'] = new TrieNode();
            iter = (iter -> children)[ch - 'a'];
        }
        iter -> terminal = true;
    }

    bool search(string word)
    {
        const TrieNode* iter = find(word);
        return iter && iter -> terminal;
    }

    bool startsWith(string prefix)
    {
        return find(prefix) != nullptr;
    }

private:
    TrieNode *root;

    void delete_sub_tree(TrieNode* node)
    {
        for(TrieNode *child: node -> children)
        {
            if(child)
                delete_sub_tree(child);
            delete child;
            child = nullptr;
        }
    }

    const TrieNode* find(const string& prefix) const
    {
        const TrieNode* iter = root;
        for(const char c: prefix)
        {
            iter = (iter -> children)[c - 'a'];
            if(iter == nullptr) break;
        }
        return iter;
    }
};

File: tree/trie/trie1.py
"""
子节点用 list 实现
"""

ALPHABETS = 26

class TrieNode:
    def __init__(self):
        self.terminal = False
        self.children = [None] * ALPHABETS

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node_iter = self.root
        for c in word:
            if node_iter.children[ord(c) - ord('a')] is None:
                node_iter.children[ord(c) - ord('a')] = TrieNode()
            node_iter = node_iter.children[ord(c) - ord('a')]
        node_iter.terminal = True

    def search(self, word: str) -> bool:
        node_iter = self._find(word)
        return node_iter is not None and node_iter.terminal

    def startsWith(self, prefix: str) -> bool:
        return self._find(prefix) is not None

    def _find(self, prefix):
        node_iter = self.root
        for c in prefix:
            node_iter = (node_iter.children)[ord(c) - ord('a')]
            if node_iter is None:
                break
        return node_iter

    def show(self):
        word = []
        self._dfs(self.root, word)

    def _dfs(self, node, word):
        if node.terminal:
            print("".join(word))
        for i in range(ALPHABETS):
            if node.children[i] is None:
                continue
            nxt = node.children[i]
            ch = chr(ord('a') + i)
            word += ch
            self.dfs(nxt, word)
            word.pop()
